packages/assets-controllers/src/DeFiPositionsController/DeFiPositionsController.ts

```diff
@@ -0,0 +1,165 @@
+  allDeFiPositions: {
+    persist: true,
+    anonymous: false,
+  },
+};
+
+export const getDefaultDefiPositionsControllerState =
+  (): DeFiPositionsControllerState => {
+    return {
+      allDeFiPositions: {},
+    };
+  };
+
+export type DeFiPositionsControllerActions =
+  DeFiPositionsControllerGetStateAction;
+
+export type DeFiPositionsControllerGetStateAction = ControllerGetStateAction<
+  typeof controllerName,
+  DeFiPositionsControllerState
+>;
+
+export type DeFiPositionsControllerEvents =
+  DeFiPositionsControllerStateChangeEvent;
+
+export type DeFiPositionsControllerStateChangeEvent =
+  ControllerStateChangeEvent<
+    typeof controllerName,
+    DeFiPositionsControllerState
+  >;
+
+/**
+ * The external actions available to the {@link DeFiPositionsController}.
+ */
+export type AllowedActions =
+  | AccountsControllerGetAccountAction
+  | AccountsControllerGetSelectedAccountAction;
+
+/**
+ * The external events available to the {@link DeFiPositionsController}.
+ */
+export type AllowedEvents =
+  | NetworkControllerStateChangeEvent
+  | AccountsControllerSelectedAccountChangeEvent;
+
+/**
+ * The messenger of the {@link DeFiPositionsController}.
+ */
+export type DeFiPositionsControllerMessenger = RestrictedMessenger<
+  typeof controllerName,
+  DeFiPositionsControllerActions | AllowedActions,
+  DeFiPositionsControllerEvents | AllowedEvents,
+  AllowedActions['type'],
+  AllowedEvents['type']
+>;
+
+/**
+ * Controller that stores assets and exposes convenience methods
+ */
+export class DeFiPositionsController extends BaseController<
+  typeof controllerName,
+  DeFiPositionsControllerState,
+  DeFiPositionsControllerMessenger
+> {
+  // TODO: Confirm whether we can store the account address instead of the id
+  // Storing the address means we don't need to query it in every event handler
+  #selectedAccountId: string;
+
+  /**
+   * Tokens controller options
+   *
+   * @param options - Constructor options.
+   * @param options.messenger - The controller messenger.
+   * @param options.state - Initial state to set on this controller.
+   */
+  constructor({
+    messenger,
+    state,
+  }: {
+    messenger: DeFiPositionsControllerMessenger;
+    state?: Partial<DeFiPositionsControllerState>;
+  }) {
+    super({
+      name: controllerName,
+      metadata: controllerMetadata,
+      messenger,
+      state: {
+        ...getDefaultDefiPositionsControllerState(),
+        ...state,
+      },
+    });
+
+    this.#selectedAccountId = this.messagingSystem.call(
+      'AccountsController:getSelectedAccount',
+    ).id;
+
+    this.messagingSystem.subscribe(
+      'AccountsController:selectedAccountChange',
+      async (selectedAccount) => {
+        this.#selectedAccountId = selectedAccount.id;
+
+        await this.#updateAccountPositions(selectedAccount.address);
+      },
+    );
+
+    this.messagingSystem.subscribe(
+      'NetworkController:stateChange',
```

Comment:
Added this subscription because we need the positions to be fetched when logging into the wallet, and sicne the `selectedAccountChange` is not triggered at the start, this seems to be the only way I found to do it.

I'd appreciate to know if there's a more idiomatic way of doing it.

Reply:
If we need to fire something when the wallet is unlocked, we can try using the keyring controller?
Example
https://github.com/MetaMask/core/blob/370147a0b78b6da9ffeafcf5bcd5b7f09d3d428c/packages/assets-controllers/src/MultichainBalancesController/MultichainBalancesController.ts#L204

Reply:
Only concern with subscribing to `:stateChange` is that this is very frequent.

If we need to use state change subscripions, consider using the 3rd parameter when creating a subscription to use a selector to reduce calls.

```
this.messagingSystem.subscribe(
  'NetworkController:stateChange', // event name
  (prevState, newState) => { ... }, // handler
  (state) => state.foo, // selector
)
```

Reply:
As per @salimtb suggestion, I have now moved towards lock and unlock keyring events. They represent the intention of what we want better and they allow us to start and stop the polling.